File: plugins/bluenimble-plugin-scripting.javascript/src/main/resources/platform/core/remote/Remote.js
function toJavaAttachments (attachments) {
	if (!attachments || !attachments.length || attachments.length <= 0) {
		return null;
	}
	var jAttachments = [];
	for (var i = 0; i < attachments.length; i++) {
		jAttachments.push (attachments [i].proxy);
	}
	return jAttachments;
}

var NoCallback = {
	onStatus: function (status, chunked, headers) {},
	onData: function (status, chunk) {},
	onDone: function (status, data) {},
	onError: function (status, message) {}
};

/**
  Represents a Remote instance<br/>
  <strong>Do not call constructor directly</strong>. 
  @see Api#remote
  @class Remote
  @classdesc 
*/

/**
  @constructor
  @access private
*/
var Remote = function (proxy) {

	/**	
	  Send a get request 
	  @param {Object} - request spec
	  @param {Object} - callback
	  @example
	  
	  api.remote (request).get ({
	  	endpoint: 'https://myserver.com',
	  	data: {
	  		email: 'john@bluenimble.com'
	  	}
	  }, {
	  	onStatus: function (status, chunked, headers) {
	  		api.logger.info (status + ': chunked: ' + chunked + ' > ' + headers);
	  	},
	  	onDone: function (status, data) {
	  		api.logger.info (status + ': data);
	  	},
	  	onError: function (status, message) {
	  		api.logger.error (status + ': ' + message);
	  	}
	  });
	*/
	this.get = function (spec, callback) {
		proxy.get (
			JC_Converters.convert (spec),
			this._callback (callback)
		);
	};
	
	/**	
	  Send a post request 
	  @param {Object} - request spec
	  @param {Object} - callback
	  @param {Array} - attachments
	  @example
	  
	  api.remote (request).post ({
	  	endpoint: 'https://myserver.com',
	  	data: {
	  		name: 'John', 
	  		age: 30, 
	  		address: '100 Rockefeller Dr, Sunnyvale CA'
	  	},
	  	headers: {
	  		'Content-Type': 'application/json'
	  	}
	  }, {
	  	onStatus: function (status, chunked, headers) {
	  		api.logger.info (status + ': chunked: ' + chunked + ' > ' + headers);
	  	},
	  	onDone: function (status, data) {
	  		api.logger.info (status + ': data);
	  	},
	  	onError: function (status, message) {
	  		api.logger.error (status + ': ' + message);
	  	}
	  });
	*/
	this.post = function (spec, callback, attachments) {
		proxy.post (
			JC_Converters.convert (spec),
			this._callback (callback),
			toJavaAttachments (attachments)
		);
	};
	
	/**	
	  Send a put request 
	  @param {Object} - request spec
	  @param {Object} - callback
	  @param {Array} - attachments
	  @example
	  
	  api.remote (request).put ({
	  	endpoint: 'https://myserver.com/123456789',
	  	data: {
	  		address: '200 Sheraton Dr, Sunnyvale CA'
	  	},
	  	headers: {
	  		'Content-Type': 'application/json'
	  	}
	  }, {
	  	onStatus: function (status, chunked, headers) {
	  		api.logger.info (status + ': chunked: ' + chunked + ' > ' + headers);
	  	},
	  	onDone: function (status, data) {
	  		api.logger.info (status + ': data);
	  	},
	  	onError: function (status, message) {
	  		api.logger.error (status + ': ' + message);
	  	}
	  });
	*/
	this.put = function (spec, callback, attachments) {
		proxy.put (
			JC_Converters.convert (spec),
			this._callback (callback),
			toJavaAttachments (attachments)
		);
	};
	
	/**	
	  Send a delete request 
	  @param {Object} - request spec
	  @param {Object} - callback
	  @param {Array} - attachments
	  @example
	  
	  api.remote (request).delete ({
	  	endpoint: 'https://myserver.com/123456789'
	  }, {
	  	onStatus: function (status, chunked, headers) {
	  		api.logger.info (status + ': chunked: ' + chunked + ' > ' + headers);
	  	},
	  	onDone: function (status, data) {
	  		api.logger.info (status + ': data);
	  	},
	  	onError: function (status, message) {
	  		api.logger.error (status + ': ' + message);
	  	}
	  });
	*/
	this.delete = function (spec, callback) {
		proxy.delete (
			JC_Converters.convert (spec),
			this._callback (callback)
		);
	};
	
	/**	
	  Send a head request 
	  @param {Object} - request spec
	  @param {Object} - callback
	  @param {Array} - attachments
	  @example
	  
	  api.remote (request).head ({
	  	endpoint: 'https://myserver.com/123456789'
	  }, {
	  	onStatus: function (status, chunked, headers) {
	  		api.logger.info (status + ': chunked: ' + chunked + ' > ' + headers);
	  	},
	  	onDone: function (status, data) {
	  		api.logger.info (status + ': data);
	  	},
	  	onError: function (status, message) {
	  		api.logger.error (status + ': ' + message);
	  	}
	  });
	*/
	this.head = function (spec, callback) {
		proxy.head (
			JC_Converters.convert (spec),
			this._callback (callback)
		);
	};
	
	/**	
	  Send a patch request 
	  @param {Object} - request spec
	  @param {Object} - callback
	  @param {Array} - attachments
	  @example
	  
	  api.remote (request).patch ({
	  	endpoint: 'https://myserver.com/123456789'
	  }, {
	  	onStatus: function (status, chunked, headers) {
	  		api.logger.info (status + ': chunked: ' + chunked + ' > ' + headers);
	  	},
	  	onDone: function (status, data) {
	  		api.logger.info (status + ': data);
	  	},
	  	onError: function (status, message) {
	  		api.logger.error (status + ': ' + message);
	  	}
	  });
	*/
	this.patch = function (spec, callback) {
		proxy.patch (
			JC_Converters.convert (spec),
			this._callback (callback)
		);
	};
	
	this._callback = function (callback) {
		if (!callback) {
			callback = NoCallback;
		}
	
		var JC_Callback = Java.extend (JC_Remote_Callback, {
			
			onStatus: function (status, chunked, headers) {
				if (!callback.onStatus) {
					return;
				}
				callback.onStatus (status, chunked, headers);
			},
			onData: function (status, chunk) {
				if (!callback.onData) {
					return;
				}
				callback.onData (status, chunk);
			},
			onDone: function (status, data) {
				if (!callback.onDone) {
					return;
				}
				callback.onDone (status, data);
			},
			onError: function (status, message) {
				if (!callback.onError) {
					return;
				}
				callback.onError (status, message);
			}
		});
		
		return new JC_Callback ();
	}
	
};

File: sdks/bluenimble-jvm-sdk/src/main/java/com/bluenimble/platform/remote/Remote.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.bluenimble.platform.remote;

import java.io.IOException;
import java.util.Map;

import com.bluenimble.platform.Feature;
import com.bluenimble.platform.Recyclable;
import com.bluenimble.platform.api.ApiStreamSource;
import com.bluenimble.platform.json.JsonObject;

@Feature (name = "remote")
public interface Remote extends Recyclable {
	
	interface Error {
		int UnknownHost = 1100;
		int Timeout 	= 1200;
		int Other 		= 5000;
	}
	
	interface Spec {
		
		String Protocol 	= "protocol";
		
		String Host			= "host";
		String Port			= "port";
		
		String Scheme 		= "scheme";
		String Endpoint 	= "endpoint";
		String Path 		= "path";
		
		String Verb 		= "verb";
		
		String Data 		= "data";
		String Headers 		= "headers";
		String Cast 		= "cast";
		String Body 		= "$body";
		
		String Ssl			= "ssl";
			String TrustAll 	= "trustAll";
			String TrustStore 	= "trustStore";
			String KeyStore 	= "keyStore";
				String StoreType 		= "type";
				String StoreAlgorithm 	= "algorithm";
				String StoreStream 		= "stream";
				String StorePassword 	= "password";
				String KeyParaphrase 	= "paraphrase";
		
		String UseStreaming	= "useStreaming";	
		String Serializer	= "serializer";
		String Sign			= "sign";
			String SignProtocol		= "protocol";
			String SignKey			= "key";
			String SignSecret		= "secret";
			String SignToken		= "token";
			String SignTokenSecret	= "tokenSecret";
			String User				= "user";
			String Password			= "password";
			String SignReplace		= "replace";
			String SignPayload		= "payload";
			String SignScheme		= "scheme";
			String SignData			= "data";
			String SignAlgorithm	= "algorithm";
			String SignHashing		= "hashing";
			String SignPlaceholder	= "signaturePlaceholder";
			String SignSignatureName= "signatureName";
			String Parameters		= "parameters";
		
		String Timeout			= "timeout";
			String TimeoutConnect	= "connect";
			String TimeoutWrite		= "write";
			String TimeoutRead		= "read";
		

		String Proxy		= "proxy";
			String ProxyType	= "type";
			String ProxyHost	= "host";
			String ProxyPort	= "port";
			
		String  Pool			= "pool";	
		
		String SuccessCode	= "successCode";
		
		String AllowProprietaryAccess = "allowProprietaryAccess";
		
	}
	
	interface Proprietary {
		String Client = "client";
	}

	interface Callback {
		void onStatus 	(int status, boolean chunked, Map<String, Object> headers);
		void onData 	(int status, byte [] chunk) 	throws IOException;
		void onDone 	(int status, Object data) 		throws IOException;
		void onError 	(int status, Object message) 	throws IOException;
	}

	Object 	proprietary (String name);
	
	void post 	(JsonObject spec, Callback callback, ApiStreamSource... attachments);
	void put 	(JsonObject spec, Callback callback, ApiStreamSource... attachments);
	void get 	(JsonObject spec, Callback callback);
	void delete (JsonObject spec, Callback callback);
	void head 	(JsonObject spec, Callback callback);
	void patch 	(JsonObject spec, Callback callback);
	
}
